pkg/scheduler/framework/framework.go

```diff
@@ -729,15 +729,87 @@ func (f *framework) runSchedulingCycleForPickNPlacementType(
+
+	// Pick the clusters.
+	//
+	// Note that at this point of the scheduling cycle, any cluster associated with a currently
+	// bound or scheduled binding should be filtered out already.
+	picked := pickTopNScoredClusters(scored, numOfClustersToPick)
+
+	// Cross-reference the newly picked clusters with obsolete bindings; find out
+	//
+	// * bindings that should be created, i.e., create a binding for every cluster that is newly picked
+	//   and does not have a binding associated with;
+	// * bindings that should be patched, i.e., associate a binding whose target cluster is picked again
+	//   in the current run with the latest score and the latest scheduling policy snapshot;
+	// * bindings that should be deleted, i.e., mark a binding as unschedulable if its target cluster is no
+	//   longer picked in the current run.
+	//
+	// Fields in the returned bindings are fulfilled and/or refreshed as applicable.
+	klog.V(2).InfoS("Cross-referencing bindings with picked clusters", "clusterSchedulingPolicySnapshot", policyRef)
+	toCreate, toDelete, toPatch, err := crossReferencePickedCustersAndObsoleteBindings(crpName, policy, picked, obsolete)
+	if err != nil {
+		klog.ErrorS(err, "Failed to cross-reference bindings with picked clusters", "clusterSchedulingPolicySnapshot", policyRef)
+		return ctrl.Result{}, err
+	}
+
+	// Manipulate bindings accordingly.
+	klog.V(2).InfoS("Manipulating bindings", "clusterSchedulingPolicySnapshot", policyRef)
+	if err := f.manipulateBindings(ctx, policy, toCreate, toDelete, toPatch); err != nil {
+		klog.ErrorS(err, "Failed to manipulate bindings", "clusterSchedulingPolicySnapshot", policyRef)
+		return ctrl.Result{}, err
+	}
+
+	// Requeue if needed.
+	//
+	// The scheduler will requeue to pick more clusters for the current policy snapshot if and only if
+	// * one or more plugins have imposed a valid batch size limit; and
+	// * the scheduler has found enough clusters for the current policy snapshot per this batch size limit.
+	//
+	// Note that the scheduler workflow at this point guarantees that the desired batch size is no less
+	// than the batch size limit, and that the number of picked clusters is no more than the batch size
+	// limit.
+	//
+	// Also note that if a requeue is needed, the scheduling decisions and condition are updated only
+	// when there are no more clusters to pick.
+	if shouldRequeue(state.desiredBatchSize, state.batchSizeLimit, len(toCreate)+len(toPatch)) {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Extract the patched bindings.
+	patched := make([]*fleetv1beta1.ClusterResourceBinding, 0, len(toPatch))
+	for _, p := range toPatch {
+		patched = append(patched, p.updated)
+	}
+
+	// Update policy snapshot status with the latest scheduling decisions and condition.
+	klog.V(2).InfoS("Updating policy snapshot status", "clusterSchedulingPolicySnapshot", policyRef)
+	if err := f.updatePolicySnapshotStatusFrom(ctx, policy, filtered, toCreate, patched, scheduled, bound); err != nil {
```

Comment:
should we check if we fully satisfy the placement requirement for the pickN?

Reply:
This change will be sent in another PR; at this moment it still follows the old logic for simplicity reasons.